### PR TITLE
feat: Have Enroll power on server and retry fetching LLDP info

### PR DESCRIPTION
**Pull request by stevekeay**
When the server is powered off the network interfaces are down, so we don't get LLDP information.

Therefore, Enroll should power on the server, then wait for LLDP information to appear before continuing.

### CODE REVIEW

File: python/understack-workflows/understack_workflows/discover.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+MIN_REQUIRED_NEIGHBOR_COUNT = 3
+
+
+def discover_chassis_info(bmc: Bmc) -> ChassisInfo:
+    """Query redfish, retrying until we get data that is acceptable.
+
+    If the server is off, power it on.
+
+    Make sure that we have at lease MIN_REQUIRED_NEIGHBOR_COUNT LLDP neighbors in
```
Comment (skrobul) [inline python/understack-workflows/understack_workflows/discover.py:19]:
```suggestion
    Make sure that we have at least MIN_REQUIRED_NEIGHBOR_COUNT LLDP neighbors in
```

File: python/understack-workflows/understack_workflows/discover.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    attempts_remaining = 6
+    while len(device_info.neighbors) < MIN_REQUIRED_NEIGHBOR_COUNT:
+        logger.info(
+            f"{bmc} does not have enough LLDP neighbors "
+            f"(saw {device_info.neighbors}), need at "
+            f"least {MIN_REQUIRED_NEIGHBOR_COUNT}. "
+        )
+        if not attempts_remaining:
+            raise Exception("No neighbors appeared")
```
Comment (skrobul) [inline python/understack-workflows/understack_workflows/discover.py:37]: This exception may be confusing/inaccurate if there the server is only partially connected
```suggestion
            raise Exception(f"{len(device_info.neighbors} out of required {MIN_REQUIRED_NEIGHBOR_COUNT}  neighbors appeared")
```

File: python/understack-workflows/understack_workflows/discover.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    the returned ChassisInfo.  If that can't be achieved in a reasonable time
+    then raise an Exception.
+    """
+    device_info = chassis_info(bmc)
+
+    if not device_info.power_on:
+        logger.info(f"Server is powered off, sending power-on command to {bmc}")
+        bmc_power_on(bmc)
+
+    attempts_remaining = 6
```
Comment (skrobul) [inline python/understack-workflows/understack_workflows/discover.py:29]: consider extracting to top-level constants

```suggestion
 LLDP_DISCOVERY_ATTEMPTS = 6
 ...
    attempts_remaining = LLDP_DISCOVERY_ATTEMPTS

Comment (stevekeay) [reply python/understack-workflows/understack_workflows/discover.py]: this is fair, done.

**Review by skrobul — APPROVED** [review]

File: python/understack-workflows/understack_workflows/main/enroll_server.py [hunk 4]
```diff
@@ -8,19 +8,23 @@
...
 from understack_workflows.bmc_settings import update_dell_drac_settings
+from understack_workflows.discover import discover_chassis_info
 from understack_workflows.helpers import credential
 from understack_workflows.helpers import parser_nautobot_args
 from understack_workflows.helpers import setup_logger
+from understack_workflows.nautobot_device import NautobotDevice
 
 logger = setup_logger(__name__)
 
+MIN_REQUIRED_NEIGHBOR_COUNT = 4
```
Comment (skrobul) [inline python/understack-workflows/understack_workflows/main/enroll_server.py:26]: sorry didn't notice this earlier - is this still used?

**Pull request merged**